src/demand.lua

```diff
@@ -119,53 +122,72 @@ function demand.getDemandFactorInfo()
 	return utils.deepCopy(DemandFactor)
 end
 
---- Updates the demand factor and returns the updated demand factor
+--- Gets the period for the timestamp
 --- @param timestamp number The current timestamp
---- @return number | nil # The demand factor, updated if necessary, nil if no update is necessary
-function demand.updateDemandFactor(timestamp)
-	if not demand.shouldUpdateDemandFactor(timestamp) then
-		return nil
-	end
-
-	local settings = demand.getSettings()
-
-	-- check that we have settings
-	if not settings then
-		return nil
-	end
+--- @return number # The period for the timestamp
+function demand.getPeriodForTimestamp(timestamp)
+	return math.floor((timestamp - demand.getSettings().periodZeroStartTimestamp) / demand.getSettings().periodLengthMs)
+		+ 1
+end
 
-	if demand.isDemandIncreasing() then
-		local upAdjustment = settings.demandFactorUpAdjustment
-		local unroundedUpdatedDemandFactor = demand.getDemandFactor() * (1 + upAdjustment)
-		local updatedDemandFactor = utils.roundToPrecision(unroundedUpdatedDemandFactor, 5)
-		demand.setDemandFactor(updatedDemandFactor)
-	else
-		if demand.getDemandFactor() > settings.demandFactorMin then
-			local downAdjustment = settings.demandFactorDownAdjustment
-			local unroundedUpdatedDemandFactor = demand.getDemandFactor() * (1 - downAdjustment)
-			local updatedDemandFactor = utils.roundToPrecision(unroundedUpdatedDemandFactor, 5)
-			demand.setDemandFactor(updatedDemandFactor)
-		end
-	end
+function demand.getTimestampForPeriod(period)
+	return demand.getSettings().periodZeroStartTimestamp + (period - 1) * demand.getSettings().periodLengthMs
+end
 
-	if demand.getDemandFactor() <= settings.demandFactorMin then
-		if demand.getConsecutivePeriodsWithMinDemandFactor() >= settings.maxPeriodsAtMinDemandFactor then
-			demand.updateFees(settings.demandFactorMin)
-			demand.setDemandFactor(settings.demandFactorBaseValue)
-			demand.resetConsecutivePeriodsWithMinimumDemandFactor()
-		else
-			demand.incrementConsecutivePeriodsWithMinDemandFactor(1)
+--- Updates the demand factor and returns the updated demand factor to the current period. If multiple periods need to be updated, this function will call itself multiple times.
+--- @param currentTimestamp number The current timestamp
+--- @return number | nil, table<number, number> # The demand factor, updated if necessary, nil if no update is necessary, and the updated demand factors
+function demand.updateDemandFactor(currentTimestamp)
+	assert(currentTimestamp, "Timestamp must be provided")
+	local settings = demand.getSettings()
+	local periodForCurrentTimestamp = demand.getPeriodForTimestamp(currentTimestamp)
+	local lastKnownPeriod = demand.getCurrentPeriod()
```

Comment:
--- Gets the current period
--- @return number # The current period
function demand.getCurrentPeriod()
	local demandFactor = utils.deepCopy(DemandFactor)
	return demandFactor and demandFactor.currentPeriod or 1
end

Nit: seems like an unnecessary deepCopy here.

Reply:
[e162b6f](https://github.com/ar-io/ar-io-network-process/pull/379/commits/e162b6f78b2652c4c398e4ef3a57f9168a45352e)